Update version value failed in non numeric semver range

`@babel/cli@^7.0.0-beta.51-beta.49` is not valid version.
It should be `@babel/cli@^7.0.0-beta.51`.

It's bug in replace semver range

https://github.com/Leko/hothouse/blob/72c24558c347c70f05c84c2a10ebd4d68396b2ec/packages/hothouse/src/Package.js#L16-L23

```
DEBUG=hothouse* hothouse -t XXX
  hothouse:cli CLI options are: { _: [],
  help: false,
  version: false,
  'per-package': true,
  p: true,
  perPackage: true,
  'dry-run': false,
  dryRun: false,
  ignore: [],
  '$0': 'hothouse',
  'package-manager': undefined,
  'repository-structure': undefined } +0ms
  hothouse:PackageManagerResolver Detect package manager in: /Users/leko/.ghq/github.com/Leko/WEB-EGG +0ms
  hothouse:PackageManagerResolver Yarn: matched=false +2ms
  hothouse:PackageManagerResolver Npm: matched=true +1ms
  hothouse:RepositoryStructureResolver Detect repository structure in: /Users/leko/.ghq/github.com/Leko/WEB-EGG +0ms
  hothouse:RepositoryStructureResolver YarnWorkspaces: matched=false +0ms
  hothouse:RepositoryStructureResolver Lerna: matched=false +0ms
  hothouse:RepositoryStructureResolver SinglePackage: matched=true +1ms
  hothouse:Engine dryRun=false +0ms
  hothouse:Engine Branch name created: hothouse-update-201863022565370 with 2018-06-30T13:56:05.370Z +1ms
  hothouse:cli Updates are: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG':
   [ { name: '@babel/cli',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: '@babel/core',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: '@babel/plugin-transform-flow-strip-types',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: '@babel/polyfill',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: false },
     { name: '@babel/preset-env',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: '@babel/preset-flow',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: '@babel/preset-react',
       current: '7.0.0-beta.49',
       latest: '7.0.0-beta.51',
       dev: true },
     { name: 'algoliasearch',
       current: '3.27.1',
       latest: '3.29.0',
       dev: false },
     { name: 'almin-logger',
       current: '6.1.5',
       latest: '6.2.1',
       dev: true },
     { name: 'babel-loader',
       current: '8.0.0-beta.3',
       latest: '7.1.4',
       dev: true },
     { name: 'classnames',
       current: '2.2.5',
       latest: '2.2.6',
       dev: false },
     { name: 'flow-bin',
       current: '0.73.0',
       latest: '0.75.0',
       dev: true },
     { name: 'react', current: '16.4.0', latest: '16.4.1', dev: false },
     { name: 'react-dom',
       current: '16.4.0',
       latest: '16.4.1',
       dev: false },
     { name: 'webpack',
       current: '4.10.2',
       latest: '4.14.0',
       dev: true } ] } +4s
  hothouse:cli UpdateChunks are: [ UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } },
  UpdateChunk {
    allUpdates: { '/Users/leko/.ghq/github.com/Leko/WEB-EGG': [Array] } } ] +0ms
  hothouse:Engine Apply update (@babel/cli 7.0.0-beta.49->7.0.0-beta.51) to WEB-EGG: { name: '@babel/cli',
  current: '7.0.0-beta.49',
  latest: '7.0.0-beta.51',
  dev: true } +4s
  hothouse:Engine Try to save: /Users/leko/.ghq/github.com/Leko/WEB-EGG/package.json +0ms
  hothouse:Engine Try to update dependencies in WEB-EGG with Npm +1ms
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but npm-shrinkwrap.json was generated for lockfileVersion@0. I'll try to do my best with it!
npm ERR! code ETARGET
npm ERR! notarget No matching version found for @babel/cli@^7.0.0-beta.51-beta.49
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
npm ERR! notarget
npm ERR! notarget It was specified as a dependency of 'webegg'
npm ERR! notarget

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/leko/.npm/_logs/2018-06-30T13_56_14_224Z-debug.log
Error: null
    at Npm.install (/usr/local/lib/node_modules/hothouse/node_modules/@hothouse/client-npm/dist/index.js:53:13)
    at SinglePackage.install (/usr/local/lib/node_modules/hothouse/dist/SinglePackage.js:19:21)
    at Engine.applyUpdates (/usr/local/lib/node_modules/hothouse/dist/Engine.js:74:38)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:678:11)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
```

--- a/packages/hothouse/__tests__/Package.spec.js
+++ b/packages/hothouse/__tests__/Package.spec.js
@@ -1,7 +1,7 @@
 /* eslint-env jest */
 // @flow
 import assert from "assert";
-import Package from "../src/Package";
+import Package, { replaceSemver } from "../src/Package";
 
 test("Package can instantiate with valid package path", () => {
   const pkg = new Package("../package.json");
@@ -9,4 +9,14 @@ test("Package can instantiate with valid package path", () => {
 });
 test("Package cannot instantiate with invalid package path", () => {
   assert.throws(() => new Package("./not-exists-path.json"));
+});
+
+test("replaceSemver can replace exact semver", () => {
+  assert.equal(replaceSemver("1.2.3", "4.5.6"), "4.5.6");
+});
+test("replaceSemver can replace tilde semver range", () => {
+  assert.equal(replaceSemver("~0.11.3", "0.12.0"), "~0.12.0");
+});
+test("replaceSemver can replace hat semver range", () => {
+  assert.equal(replaceSemver("^1.13.1", "2.0.0"), "^2.0.0");
 });

--- a/packages/hothouse/__tests__/Package.spec.js
+++ b/packages/hothouse/__tests__/Package.spec.js
@@ -20,3 +20,12 @@ test("replaceSemver can replace tilde semver range", () => {
 test("replaceSemver can replace hat semver range", () => {
   assert.equal(replaceSemver("^1.13.1", "2.0.0"), "^2.0.0");
 });
+test("replaceSemver can replace semver range with prerelease (like Babel)", () => {
+  assert.equal(
+    replaceSemver("^7.0.0-beta.49", "7.0.0-beta.51"),
+    "^7.0.0-beta.51"
+  );
+});
+test("replaceSemver throws Error when too many diffs", () => {
+  assert.throws(() => replaceSemver("^1.0.0-beta.1", "1.0.0-beta.202"));
+});

--- a/packages/hothouse/src/Package.js
+++ b/packages/hothouse/src/Package.js
@@ -3,6 +3,14 @@ import fs from "fs";
 import semver from "semver";
 import type { Update } from "@hothouse/types";
 
+export const replaceSemver = (
+  fromVersion: string,
+  toVersion: string
+): string => {
+  const current = semver.coerce(fromVersion).version;
+  return fromVersion.replace(current, toVersion);
+};
+
 export default class Package {
   pkgJsonPath: string;
   pkgJson: Object;
@@ -17,9 +25,8 @@ export default class Package {
     const deps = update.dev
       ? this.pkgJson.devDependencies
       : this.pkgJson.dependencies;
-    const current = semver.coerce(deps[update.name]).version;
 
-    deps[update.name] = deps[update.name].replace(current, update.latest);
+    deps[update.name] = replaceSemver(deps[update.name], update.latest);
   }
 
   async save(): Promise<void> {

--- a/packages/hothouse/src/Package.js
+++ b/packages/hothouse/src/Package.js
@@ -4,11 +4,30 @@ import semver from "semver";
 import type { Update } from "@hothouse/types";
 
 export const replaceSemver = (
-  fromVersion: string,
+  fromVersionRange: string,
   toVersion: string
 ): string => {
-  const current = semver.coerce(fromVersion).version;
-  return fromVersion.replace(current, toVersion);
+  const UPDATE_LIMITS = 200;
+  const prefixRegEx = /^(\^|~|>=|>|<|<=)?\s*(.*)$/;
+  if (!prefixRegEx.test(fromVersionRange)) {
+    throw new Error(`Unsupported semver format: ${fromVersionRange}`);
+  }
+  // $FlowFixMe(already-tested)
+  const [, prefix, fromVersion] = fromVersionRange.match(prefixRegEx);
+  let updated = fromVersion;
+  for (let i = 0; i < UPDATE_LIMITS; i++) {
+    const diff = semver.diff(updated, toVersion);
+    if (!diff) {
+      break;
+    }
+    updated = semver.inc(updated, diff);
+  }
+  if (semver.diff(updated, toVersion)) {
+    throw new Error(
+      `Too many diffs to update to ${toVersion} from ${fromVersionRange}`
+    );
+  }
+  return (prefix || "") + updated;
 };
 
 export default class Package {

